### PR TITLE
sqlite3: refactor: recursive traversal; cleanups 

### PR DESCRIPTION
**Pull request by zellyn**
Major change:

- Add `visitRawInorder` and `visitRecordsInorder` to btree, which will
  load and traverse child pages recursively, and use them instead of
  iterating.

Smaller cleanups:

- Stash the db, not the dbheader in `btreeTable`, so we can decode
  child pages.
- Rename `load` to `decodeRecord`, and make it take just the payload
  bytes.
- Create `loadCell`, a wrapper around `parseCell` that first seeks to
  the right place. All callers used to do it themselves.
- Remove unused arg from `parseCell`, to emphasize that it only parses
  the cell at the current seek position.
- Remove unused struct type `btreeInteriorTable`.
- Rename `cellInfo.Key` to `cellInfo.LeftChildPage` -- as far as I can
  tell, no int32 "Key" is mentioned in the data format description --
  and don't put sizes in it.
- Make `RowID` in `cellInfo` a pointer, since it's optional, but may
  be zero.

### CODE REVIEW

**Comment by zellyn** [comment]
I realize this is a big change with lots of things mixed together. On reflection, I'm probably too used to our work workflow, where you squash all commits together before merging. Let me know if I should break the changes up into multiple commits.

**Comment by sbinet** [comment]
nah... it's ok, although I would have perhaps split it in 2: one CL with the recursive btree traversal and another one with the cleanups.

File: cell.go [hunk 1]
```diff
@@ -6,8 +6,8 @@ package sqlite3
 
 // cellInfo holds information about an on-disk cell.
 type cellInfo struct {
-	Key          int64
-	RowID        int64
-	Payload      []byte
-	OverflowPage int32
+	LeftChildPage int32
+	RowID         *int64
```
Comment (sbinet) [inline cell.go]: why the change to a `*int64` ?
could you explain this a bit?

Comment (zellyn) [reply cell.go]: Not all btrees have rowids, so it's optional. But I think zero is a valid rowid. Actually, I'll check that...

Comment (zellyn) [reply cell.go]: ... oh yeah, unique integer columns alias to rowid, so zero is definitely allowed.

So you need a nil pointer to signal no rowid

Comment (sbinet) [reply cell.go]: what about using -1 instead ?

Comment (zellyn) [reply cell.go]: > The rowid of the SQL table is the 64-bit signed integer key for each entry in the table b-tree.

-1 is a valid value for RowIDs. There's no available, invalid value to signal a missing RowID.

Comment (sbinet) [reply cell.go]: ok. I tried to reduce allocs :)

**Comment by sbinet** [comment]
could you rebase and resolve the conflicts?
(apologies for leaving this lingering a bit)

**Comment by zellyn** [comment]
> could you rebase and resolve the conflicts?
Done.

> (apologies for leaving this lingering a bit)
On the contrary, you've been remarkably responsive :-)
